### PR TITLE
fix: Fixes storage behavior

### LINKED ISSUE #852: Change Store Iterator method to accept startKey and endKey

### LINKED ISSUE #1946: indexeddb and mem have different behavior
We have several implementations of the storage.
couchdb,leveldb and mysql have the same behavior. But indexeddb and mem  are not. 
We need to make sure that all implementations have the same behavior.

### PR DESCRIPTION
**Pull request by soluchok**
Closes #1946 
Closes #852 

Signed-off-by: Andrii Soluk <isoluchok@gmail.com>

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -83,6 +83,9 @@ linters-settings:
 linters:
   enable-all: true
   disable:
+    - godot
+    - nestif
+    - testpackage
```
Comment (llorllale) [inline .golangci.yml:88]: why are we disabling these?

**Comment by llorllale** [comment]
@soluchok it seems like the test cases are duplicated among the stores. We should have all test cases in one place and subject all implementations to those.

File: .golangci.yml (hunk 1 above)
Comment (soluchok) [reply .golangci.yml:88]: I have updated golangci to support t.Cleanup function (added in go 1.4). These linters were added in the new version.

**Comment by soluchok** [comment]
> @soluchok it seems like the test cases are duplicated among the stores. We should have all test cases in one place and subject all implementations to those.

I moved common tests to `pkg/storage/store_test.go`. If I remove duplicates it will decrease test coverage. The idea is to keep common tests in `pkg/storage`. (responsible for behavior). And each storage implementation should have its own unit tests.

File: pkg/storage/couchdb/couchdbstore.go [hunk 2]
```diff
@@ -155,12 +155,7 @@ func (c *CouchDBStore) Put(k string, v []byte) error {
 		return errors.New("key and value are mandatory")
 	}
 
-	var valueToPut []byte
-	if isJSON(v) {
-		valueToPut = v
-	} else {
-		valueToPut = wrapTextAsCouchDBAttachment(v)
-	}
+	valueToPut := wrapTextAsCouchDBAttachment(v)
```
Comment (fqutishat) [inline pkg/storage/couchdb/couchdbstore.go:158]: we shouldn't save data in Attachment it's bad for performance. we should check if data format is json will save in fields

Comment (fqutishat) [reply pkg/storage/couchdb/couchdbstore.go:158]: you delete the correct behaviour

Comment (soluchok) [reply pkg/storage/couchdb/couchdbstore.go:158]: in that case, implementation is incompatible since we cannot use keys such as "_key, _rev, etc."

Comment (soluchok) [reply pkg/storage/couchdb/couchdbstore.go:158]: it looks like one implementation allows you to do that and another does not.

File: pkg/storage/mem/mem_store.go [hunk 3]
```diff
@@ -127,19 +128,49 @@ func (s *memStore) Get(k string) ([]byte, error) {
 
 // Iterator returns iterator for the latest snapshot of the underlying db.
 func (s *memStore) Iterator(start, limit string) storage.StoreIterator {
-	// TODO Change Store Iterator https://github.com/hyperledger/aries-framework-go/issues/852
+	if start == "" {
```
Comment (ToJen) [inline pkg/storage/mem/mem_store.go:131]: just wondering if `limit` may also need to be checked

File: pkg/storage/couchdb/couchdbstore.go [hunk 4]
```diff
@@ -155,38 +154,30 @@ func (c *CouchDBStore) Put(k string, v []byte) error {
...
-	} else {
-		valueToPut = wrapTextAsCouchDBAttachment(v)
-	}
-
 	revID, err := c.getRevID(k)
 	if err != nil {
 		return err
 	}
 
+	doc := map[string]interface{}{"payload": v}
```
Comment (fqutishat) [inline pkg/storage/couchdb/couchdbstore.go:162]: you can put all data in one field

Comment (fqutishat) [reply pkg/storage/couchdb/couchdbstore.go:162]: you need to check data type

File: pkg/storage/couchdb/couchdbstore.go [hunk 5]
```diff
@@ -155,45 +154,30 @@ func (c *CouchDBStore) Put(k string, v []byte) error {
 		return errors.New("key and value are mandatory")
 	}
 
-	var valueToPut []byte
```
Comment (fqutishat) [inline pkg/storage/couchdb/couchdbstore.go:158]: we need this logic

File: pkg/storage/couchdb/couchdbstore.go (hunk 2 above)
Comment (soluchok) [reply pkg/storage/couchdb/couchdbstore.go]: @fqutishat  I have changed the logic, review it again

File: pkg/storage/couchdb/couchdbstore.go (hunk 4 above)
Comment (soluchok) [reply pkg/storage/couchdb/couchdbstore.go:162]: we do not need to, it is always `[]byte`

Comment (soluchok) [reply pkg/storage/couchdb/couchdbstore.go:162]: @fqutishat  why I cannot put data to one field? it is always `[]byte` I put `[]byte` I get `[]byte`.
It solves the problem and it even improves performance (less Marshall/Unmarshall).

Comment (fqutishat) [reply pkg/storage/couchdb/couchdbstore.go:162]: it's not good if data is big

Comment (fqutishat) [reply pkg/storage/couchdb/couchdbstore.go:162]: in some cases you need to save big binary in attachment like images

Comment (fqutishat) [reply pkg/storage/couchdb/couchdbstore.go:162]: @troyronda

Comment (soluchok) [reply pkg/storage/couchdb/couchdbstore.go:162]: ok, I will combine it

Comment (llorllale) [reply pkg/storage/couchdb/couchdbstore.go:162]: @soluchok please put a comment in there with the rationale, much appreciated :)

Comment (soluchok) [reply pkg/storage/couchdb/couchdbstore.go]: I rollbacked changes. But I wrapped JSON payload with the `payload` field.
@llorllale the problem was that we couldn't use keys such as "_key, _rev, etc."

File: pkg/storage/mem/mem_store.go (hunk 3 above)
Comment (soluchok) [reply pkg/storage/mem/mem_store.go]: Fixed. It supposed to be a `limit` not a `start`. Added common tests for this logic.

File: .golangci.yml (hunk 1 above)
Comment (troyronda) [reply .golangci.yml:88]: Can we add an issue and comment to followup on these news linters? I haven't looked into what they analyze.

Comment (soluchok) [reply .golangci.yml:88]: Sure, looks like we need `nestif` (for sure). We do not need `testpackage`(for sure). And probably we need `godot`.
godot is about to add `.` at the end of comments.

Comment (soluchok) [reply .golangci.yml:88]: @troyronda FYI: it is not the latest version of `golangci.` There are even more linters. I will create an issue to update the  `golangci` to the latest version.

Comment (troyronda) [reply .golangci.yml:88]: Sounds good. @soluchok

Comment (soluchok) [reply .golangci.yml:88]: I've created an issue https://github.com/hyperledger/aries-framework-go/issues/2032

**Pull request merged**